File: Competitive Programming 3/Section 1.4/Real Life Harder/538 - Balancing Bank Accounts/Java/Main.java
import java.util.*;

public class Main{
    static Scanner sc = new Scanner(System.in);
    static class S{
        public String name;
        public int amount;
        public S(String name,int amount){
            this.name=name;
            this.amount=amount;
        }
    }
    public static void main(String[]args){
        int n,t;
        for(int cs=1;(n=sc.nextInt())!=0;cs++){
            t = sc.nextInt();
            System.out.printf("Case #%d\n",cs);
            HashMap<String,Integer> balance = new HashMap<String,Integer>();
            for(int i = 0 ; i < n ; i++)
                balance.put(sc.next(),0);
            for(int i = 0 ; i < t ; i ++){
                String n1 = sc.next(), n2 = sc.next();
                int amount = sc.nextInt();
                balance.put(n1,balance.get(n1)-amount);
                balance.put(n2,balance.get(n2)+amount);
            }
            Vector<S> positives = new Vector<S>(), negatives = new Vector<S>();
            int iP = 0 , iN = 0 ;
            for(Map.Entry b : balance.entrySet()){
                String name = (String)b.getKey();
                int amount = (int)b.getValue();
                if (amount > 0)
                    positives.add(new S(name,amount));
                else if(amount < 0)
                    negatives.add(new S(name,-amount));
                String aS,bS;
                amount = 0 ;
                while(iP<positives.size()&&iN<negatives.size()){
                    aS=positives.get(iP).name;
                    bS=negatives.get(iN).name;
                    if(positives.get(iP).amount<negatives.get(iN).amount){
                        negatives.get(iN).amount-= amount = positives.get(iP++).amount;
                    }else if(positives.get(iP).amount>negatives.get(iN).amount){
                        positives.get(iP).amount-= amount = negatives.get(iN++).amount;
                    }else{
                        amount=positives.get(iP).amount;
                        iN++;
                        iP++;
                    }
                    System.out.printf("%s %s %d\n",aS,bS,amount);
                }
            }
            System.out.println(); 
        }
    }
}

File: Competitive Programming 3/Section 1.4/Real Life Harder/538 - Balancing Bank Accounts/C++/main.cpp
#include<bits/stdc++.h>

using namespace std;

int n,t;
string line;

int main(){
    ios_base::sync_with_stdio(0);cin.tie(0);
    for(int cs=1;cin>>n>>t,n;cs++){
        printf("Case #%d\n",cs);
        map<string,int> balance;
        while(n--){
            cin >> line;
            balance[line];
        }
        while(t--){
            string n1,n2;
            int amount;
            cin >> n1 >> n2 >> amount;
            balance[n1]-=amount;
            balance[n2]+=amount;
        }
        vector<pair<string,int> > positives,negatives;
        int iP=0,iN=0;
        for(auto b : balance){
            string name = b.first;
            int amount = b.second;
            if(amount>0)
                positives.push_back(make_pair(name,amount));
            else if(amount<0)
                negatives.push_back(make_pair(name,-amount));
            string aS,bS;
            amount = 0 ;
            for(;iP<positives.size()&&iN<negatives.size();iN++,iP++){
                aS = positives[iP].first;
                bS = negatives[iN].first;
                if(positives[iP].second<negatives[iN].second)
                    negatives[iN--].second -= amount = positives[iP].second;
                else if(positives[iP].second>negatives[iN].second)
                    positives[iP--].second -= amount = negatives[iN].second;
                else
                    amount = positives[iP].second;
                printf("%s %s %d\n",aS.c_str(),bS.c_str(),amount);
            }
        }
        printf("\n");
    }
    return 0;
}